add tool for pushing vendor updates to consuming repos easier

dhellmann opened the pull request.

dhellmann commented:
/cc @elmiko @JoelSpeed

Danil-Grigorev commented on hack/push_updates.sh:
```diff
@@ -0,0 +1,140 @@
+EOF
+}
+
+DEFAULT_REPOS="
+cluster-api-provider-baremetal
+cluster-api-provider-gcp
+"
```
These are missing multiple repos. 
The full list should contain:
- cluster-api-provider-gcp
- cluster-api-provider-aws
- cluster-api-provider-azure
- cluster-api-provider-openstack
- cluster-api-provider-baremetal
- cluster-api-provider-ovirt

dhellmann replied:
Added

Danil-Grigorev commented on hack/push_updates.sh:
```diff
@@ -0,0 +1,140 @@
+if [ -z "$SHA" ]; then
+    echo "Determining SHA for machine-api-operator repository..."
+    hub clone openshift/machine-api-operator
+    (cd ./machine-api-operator && git checkout origin/$BRANCH_NAME)
```
I wonder if this should not be done in a temporary directory, and additionally cleaned up after completion?

dhellmann replied:
It is a temporary directory, see lines 86-87.

I didn't add a trap to clean it up because @elmiko wanted it to run in a container, and I figured everything would be removed when the container is removed. Maybe that assumption isn't safe or some folks will run this outside of the container?

elmiko replied:
i certainly don't want to see us do extra work just for the container option. that said, it should be relatively safe to leave artifacts in a temporary directory as it will get cleaned by the o/s as needed. additionally it might help for debugging to have the artifacts.

dhellmann replied:
It certainly helped while developing the script. :-)

Danil-Grigorev commented on hack/push_updates.sh:
```diff
@@ -0,0 +1,140 @@
+# Push updates that need to be vendored in other repositories.
+
+if ! which hub 2>/dev/null 1>&2; then
+    echo "This tool requires the hub command line app." 1>&2
```
Is it possible just to use `git` tool?.. Don't know all the details, but from visual inspection, it is possible to replicate all commandls with `git clone` and `git remote add ...`

dhellmann replied:
git can't create the fork if you don't already have it

dhellmann commented:
/retest

JoelSpeed commented on hack/push_updates.sh:
```diff
@@ -0,0 +1,144 @@
+    hub clone openshift/$repo
+    pushd ./$repo
+
+    hub fork --remote-name $REMOTE_NAME
```
Does this succeed if there is already a fork?

dhellmann replied:
Yes, it recognizes the fork even if it has a different name and sets up the remote appropriately.

dhellmann commented on hack/push_updates.sh:
```diff
@@ -0,0 +1,144 @@
+
+cat <<EOF
+
+  -s SHA
+    The SHA from machine-api-operator repository to push.
+    Defaults to HEAD of master.
```
It's actually any reference, but it seemed safer to suggest using a SHA because if you just pass "HEAD" or a branch name then the commit message wouldn't be very meaningful. The short SHA should be OK.

JoelSpeed commented:
I think I'm happy with this as is now, would be good to see it in action (maybe dry run?) so we can verify it works before merging, what do people think?

dhellmann commented:
> I think I'm happy with this as is now, would be good to see it in action (maybe dry run?) so we can verify it works before merging, what do people think?

Sure, that makes sense. Do you want me to do that and link to some PRs, or do you want to give it a spin to see how it behaves first hand?

/retest

JoelSpeed commented:
Just cloned this myself to have a play, not sure how important these are but some observations:
- You can't get the help text without hub installed
- The name of the script uses an underscore, our convention for this is hyphens (all bar 1 other are hyphens)
- Using dry run, I got an error `Error creating fork: JoelSpeed/cluster-api-provider-azure already exists on github.com` which I wasn't expecting to see, this caused the script to exit. My CAPZ fork is directly from kubernetes-sigs, not from the openshift fork, I wonder how many people will have that problem or if there is a workaround 🤔 
- Commit message was `Update machine-api-operator to 8f5ef9` for AWS, looks good, had intended to have 7 chars but my suggestion should have been `${SHA:0:7}` instead for 7 characters, not sure how many we want? 
- The AWS clone, update commit all seemed to work well :)

dhellmann commented:
> Just cloned this myself to have a play, not sure how important these are but some observations:
> 
> * You can't get the help text without hub installed

Fixed

> * The name of the script uses an underscore, our convention for this is hyphens (all bar 1 other are hyphens)

Fixed

> * Using dry run, I got an error `Error creating fork: JoelSpeed/cluster-api-provider-azure already exists on github.com` which I wasn't expecting to see, this caused the script to exit. My CAPZ fork is directly from kubernetes-sigs, not from the openshift fork, I wonder how many people will have that problem or if there is a workaround 🤔

I changed the logic so the fork isn't created in dry-run mode.

I _think_ if you create a fork yourself using a different repository name the hub command will use the Github API to figure that out. I could also see if I can make it try a simple clone if the fork fails.

> * Commit message was `Update machine-api-operator to 8f5ef9` for AWS, looks good, had intended to have 7 chars but my suggestion should have been `${SHA:0:7}` instead for 7 characters, not sure how many we want?

I switched to using the abbreviated hash computed by git instead of truncating it in the script.

> * The AWS clone, update commit all seemed to work well :)

dhellmann commented:
> I _think_ if you create a fork yourself using a different repository name the hub command will use the Github API to figure that out. I could also see if I can make it try a simple clone if the fork fails.

On second thought, I'm not sure that's actually possible.

I wonder if because of the error handling we would actually want the script to run in 2 different phases. First to build all of the patches, and then either a second run to submit them or just leave it up to the user to submit them (and automate submission separately when we put this into a prow job).

Thoughts?

JoelSpeed commented on hack/push-updates.sh:
```diff
@@ -0,0 +1,161 @@
+
+    # Use -d option to avoid build errors because we aren't going to
+    # build locally.
+    go get -d github.com/openshift/machine-api-operator@${FULL_SHA}
```
I don't think FULL_SHA is initialised when SHA is set explicitly, looks like it's only set when SHA is empty and we work it out in the script?

dhellmann replied:
I fixed that, and also split the operation into 2 loops as we discussed in the meeting this morning.

JoelSpeed reviewed (commented):
> I switched to using the abbreviated hash computed by git instead of truncating it in the script.

+1 for this

I think the fallback to simple clone looks good, I haven't tested it, but I think it should work. I personally would prefer to keep the script so that it automates the whole thing. If I have to delete some of my forks and have them properly forked from the openshift forks to make that work, I think that's a sacrifice I'd be willing to make.

JoelSpeed commented:
@dhellmann Used the script earlier and it pretty much worked a treat! (https://bugzilla.redhat.com/show_bug.cgi?id=1877743) Had to manually fix one up because they're using a `replace` for one of the dependencies, and I had to set `github.user` in my git config, but then the fallback to simple clone worked as well.

Only thing I will say is that it doens't look like ovirt is in a place where they need this script just yet, might be worth dropping them from the defaults for now. 

From this experience, future improvement ideas:
- Reuse an existing directory: In case of failure, could re-use existing cloned resources? 
- Allow a PR description to be provided
- Pause on commit failure to allow you to fix up, then continue?

dhellmann commented:
> @dhellmann Used the script earlier and it pretty much worked a treat! (https://bugzilla.redhat.com/show_bug.cgi?id=1877743) Had to manually fix one up because they're using a `replace` for one of the dependencies, and I had to set `github.user` in my git config, but then the fallback to simple clone worked as well.

I think I assumed the GitHub.user setting was part of configuring hub, but apparently not. It looks like the script could read ~/.config/hub instead. At the very least we could have it require the user to set up the git config it's expecting.

> Only thing I will say is that it doens't look like ovirt is in a place where they need this script just yet, might be worth dropping them from the defaults for now.

Makes sense.

> From this experience, future improvement ideas:
> 
> * Reuse an existing directory: In case of failure, could re-use existing cloned resources?

I thought about making it create the working directory with the SHA in the name, so it could find it again later. Then it could skip building the patch if it sees a cloned repo in that predictable temp directory, assuming that the user would have fixed whatever caused the failure before running again. What do you think?

> * Allow a PR description to be provided

Definitely.

> * Pause on commit failure to allow you to fix up, then continue?

See above

dhellmann commented:
I rebased, squashed, and removed ovirt from the list of repositories.

dhellmann commented:
/retest

JoelSpeed commented:
/approve

Thanks @dhellmann

elmiko commented on hack/push-updates.sh:
```diff
@@ -0,0 +1,174 @@
+
+function fork_repo () {
+    local repo_name="$1"
+    local remote_name="$2"
```
it seems like `remote_name` is unused?

dhellmann replied:
d'oh, it's using REMOTE_NAME below instead of the local variable.

elmiko replied:
i figured it had to be one or the other XD

dhellmann replied:
I removed the extra variable.

elmiko reviewed (commented):
looks mostly good to me, just had one question

elmiko reviewed (approved):
/lgtm

JoelSpeed commented:
/hold

This will just keep retesting unnecessarily until it's merged

Do we want to get this merged before master opens or do we not really care?

dhellmann commented:
It doesn't really make any difference to me, so if you'd like to wait that's fine.

JoelSpeed commented:
Saves us having to create a BZ, so let's just hold cancel when master opens, we can all use it from here until then

elmiko commented:
just wanted to note that i tried the script out in dry run mode today and everything is working as i would expect.

JoelSpeed commented:
/hold cancel

I think this should be able to merge now that the 4.6 branch has been cut 🤔

elmiko commented:
/lgtm
